[WOO POS] create and sync order, collect payment

bozidarsevo opened the pull request:
<!-- Remember about a good descriptive title. -->

#12763 #13021 #12983
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

- Updated PointOfSaleDashboardViewModel to include logic for creating and syncing of the order
- created order will be used for collecting the payment
- processing the payment
- UI will be finalized in one of the next PRs alogside with better order cleanup for next transaction and payment processing

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

- Enter POS experience
- Add few products in the cart
- Tap "Checkout"
- Check that the totals and payments screen is shown with totals amounts being calculated and shown
- Use the card reader to collect the payment
- Check that the payment is successful and "New transaction" button is shown

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

![order payment](https://github.com/woocommerce/woocommerce-ios/assets/6242034/b021844e-83f3-40d9-81e9-a0f3c1dbd8ce)

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

bozidarsevo committed:
Update POS totals calculation with order creation and sync logic

bozidarsevo committed:
Update TotalsView.swift

bozidarsevo committed:
Merge branch 'trunk' into feat/12763-pos-order-totals-order-creation-sync

bozidarsevo committed:
Resolved merge conflicts

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

bozidarsevo commented on Yosemite/Yosemite/PointOfSale/POSItemProvider.swift:
```diff
@@ -3,6 +3,7 @@ public protocol POSItem {
     var productID: Int64 { get }
     var name: String { get }
     var price: String { get }
+    var formattedPrice: String { get }
```
Added formattedPrice because clean price is expected when creating/syncing the order to get a proper totals calculation

joshheald replied:
Is this for a local calculation before the Order comes back with the server-side calculation? I thought we were going to rely on the backend for all the numbers in MVP... and show shimmering boxes while they're loading. Perhaps there's been more discussion since then, or I'm misunderstanding it, though.

bozidarsevo replied:
formatted price is a price with the currency symbol.. because the price coming from backend is without the currency symbol. So to avoid doing the currency formatting all the time I just added this to be available

joshheald replied:
Sure, that makes sense. 

My question was about why we still need `price` though? I didn't think we were going to do any calculations in the app for MVP...

bozidarsevo committed:
Extract createInitialOrder to private extension

bozidarsevo committed:
Update PointOfSaleDashboardViewModel.swift

bozidarsevo committed:
Remove checkoutButtonDisabled

bozidarsevo committed:
Update PointOfSaleDashboardViewModel.swift

bozidarsevo committed:
Use formattedPrice

bozidarsevo committed:
Temporary call for collecting the payment

bozidarsevo commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -208,46 +271,55 @@ private extension PointOfSaleDashboardViewModel {
+                                                              allProducts: self.items)
             self.order = order
-            return order
+            // TODO: this is temporary solution
```
This is just here to trigger payment temporary. We will have a proper solution in next PRs

joshheald replied:
The intention with the async task here was that if called from SwiftUI directly, it would be cancelled when the view went offscreen. It's fine if we can't do that because of the dependency on creating/updating the order, but we'll still need to make sure that cancel gets called.

bozidarsevo commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -90,45 +117,52 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
 
         // TODO: Here we should present something to show the payment was successful or not,
         // and then clear the screen ready for the next transaction.
+        switch paymentResult {
```
@joshheald do you know why this switch is not being executed? Is `cardPresentPaymentService.collectPayment` working ok and returning a value? Maybe `paymentTask` is buggy

joshheald replied:
Yep there's a bug in the service where it never calls completion.

The switch works fine for cancellation, but the success case doesn't get called.

There's a couple of options to fix it, I'll get that done today.

joshheald replied:
Fixed in bc8aaa7

bozidarsevo committed:
Move collect payment button

joshheald committed:
Return card payment result when payment finishes

Previously, calls to `CardPresentPaymentCollectOrderPaymentUseCaseAdaptor().collectPaymentTask` would never return if the payment was successful, only if it was cancelled.

The completion handler we gave the `CollectOrderPaymentUseCase` wasn’t being called, because it was passed as the overall usecase’s completion. That closure is only called when the merchant dismisses or discards the receipt, in the existing modal, which marks the very end of the modal flow.

We should have actually used the `onPaymentCompletion` handler, which is called at the end of the `attemptPayment` completion handler, after a successful payment. Failed or cancelled payments return earlier, so we can safely say that any `onPaymentCompletion` call comes after a successful payment.

This commit fixes the issue by moving the contents of our adaptor’s `onCompleted` handler to the `onPaymentCompleted` handler instead. I also removed the `.idle` event, which had the result of preventing the `Payment successful` message from ever being shown.

joshheald committed:
Merge branch 'feat/12763-pos-order-totals-order-creation-sync' of github.com:woocommerce/woocommerce-ios into feat/12763-pos-order-totals-order-creation-sync

bozidarsevo committed:
Add shimmering

bozidarsevo committed:
Update TotalsView.swift

bozidarsevo committed:
Update divider and gradient

joshheald commented on WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift:
```diff
@@ -17,6 +17,7 @@ struct PointOfSaleCardPresentPaymentInLineMessage: View {
             PointOfSaleCardPresentPaymentTapSwipeInsertCardMessageView(viewModel: viewModel)
         case .processing:
             Text("processing...")
+                .shimmering()
```
Is this meant to shimmer? It seems a bit odd. Doesn't matter much if it's just temporary to give the impression it's doing something, this view can be properly built out later

bozidarsevo replied:
Yeah, I put it as a temp, will remove it!

joshheald commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -100,8 +100,12 @@ private extension CartView {
-                                                      cardPresentPaymentService: CardPresentPaymentPreviewService()))
+                                                      cardPresentPaymentService: CardPresentPaymentPreviewService(),
+                                                      orderService: POSOrderService(siteID: Int64.min,
+                                                                                    credentials: Credentials(authToken: "token"))))
```
Perhaps make an order preview service here, so we don't need these extra imports or verbose preview set up?

joshheald commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenu.swift:
```diff
@@ -168,7 +168,9 @@ private extension HubMenu {
-                        cardPresentPaymentService: cardPresentPaymentService)
+                        cardPresentPaymentService: cardPresentPaymentService,
+                        orderService: POSOrderService(siteID: viewModel.siteID,
+                                                      credentials: ServiceLocator.stores.sessionManager.defaultCredentials!))
```
It shouldn't fail, but it's probably better if we avoid this force unwrap.

joshheald replied:
I've sorted this in 4443bf3, let me know if you have any concerns

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -208,46 +265,55 @@ private extension PointOfSaleDashboardViewModel {
-                                                               taxable: false,
-                                                               orderNote: order.customerNote,
-                                                               email: "") { result in
+                                                               taxable: false) { result in
```
Was the update function removed because of a merge thing? Perhaps we can just delete all the test order code now?

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -208,46 +265,55 @@ private extension PointOfSaleDashboardViewModel {
+            Task {
+                await self.totalsViewWillAppear()
+            }
+            print("🟢 [POS] Synced order: \(order)")
```
Consider `DDLogInfo` if we're going to merge these... we don't usually keep print statements in the code, as we want things we leave in to go to the device logs.

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -208,46 +265,55 @@ private extension PointOfSaleDashboardViewModel {
+            self.isSyncingOrder = true
+            let order = try await self.orderService.syncOrder(cart: cart,
+                                                              order: self.order,
+                                                              allProducts: self.items)
```
```suggestion
            isSyncingOrder = false
        }
        do {
            isSyncingOrder = true
            let order = try await orderService.syncOrder(cart: cart,
                                                         order: order,
                                                         allProducts: items)
```

No need for `self` here

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -208,46 +265,55 @@ private extension PointOfSaleDashboardViewModel {
+            // TODO: this is temporary solution
+            Task {
+                await self.totalsViewWillAppear()
+            }
```
```suggestion
            await totalsViewWillAppear()
```

No need for self, and we're in an `async` function so no need to wrap this in a Task that I can see...

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -90,69 +115,101 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
+        itemsInCart.removeAll()
+        orderStage = .building
+        paymentState = .acceptingCard
+        order = nil
```
I think that when this happens after a successful card payment, we probably need to find a way to call the completion handler from the card payment service. Not totally essential but it's a good way to reset the service, and make sure that any cleanup we need happens.

If not, then a `reset` function might need called.

These are just thoughts/notes for me, nothing to change on this PR!

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -90,69 +115,101 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
+        guard let order,
+              cardReaderConnectionViewModel.connectionStatus == .connected else {
+            return
+        }
```
The designs have a `Collect Payment` button on them when the reader is disconnected – it's gone with this PR, but we should add it back.

It should only show up when the reader isn't connected, but if the user taps it, it's totally fine for us to call `collectPayment` and for it to do its work. It'll start the connection process automatically, which is what we want when it's called for a user-initiated action. 

We just don't want to call it with a disconnected reader when we land on the totals screen, because that would mean immediately putting a connection dialogue in their way.

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -63,11 +80,19 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
+        }
     }
 
     func removeItemFromCart(_ cartItem: CartItem) {
         itemsInCart.removeAll(where: { $0.id == cartItem.id })
         checkIfCartEmpty()
+
+        if orderStage == .finalizing {
+            startSyncingOrder()
+        }
```
I think with the designs/decisions we've taken, we want this to be impossible when the order's being finalised, right? It's fine to be permissive here, but we'll need to make sure we have the same lifecycle for the payment intent when it's been created. Perhaps it would be better to prevent this from working from the checkout screens?

bozidarsevo replied:
Yep, for now we still show remove buttons when in checkout, so when we remove those I will remove this also

joshheald commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -138,40 +126,51 @@ private extension TotalsView {
-            Button(action: viewModel.cardPaymentTapped) {
-                Text("Collect Payment")
-            }
+            Text("Reader disconnected")
```
I think we want this button back...

joshheald commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -138,40 +126,51 @@ private extension TotalsView {
                 PointOfSaleCardPresentPaymentInLineMessage(messageType: inlinePaymentMessage)
             } else {
                 Text("Reader connected")
+                Button(action: viewModel.cardPaymentTapped) {
+                    Text("Collect Payment")
+                }
```
If the reader's connected, we don't necessarily need this button, because the payment should start automatically when the view is shown.

I can see it being useful after a payment's been cancelled though...

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -100,11 +100,15 @@ fileprivate extension CardPresentPaymentEvent {
 
 #if DEBUG
+import class Yosemite.POSOrderService
+import enum Yosemite.Credentials
 #Preview {
     NavigationStack {
         PointOfSaleDashboardView(
             viewModel: PointOfSaleDashboardViewModel(items: POSItemProviderPreview().providePointOfSaleItems(),
-                                                     cardPresentPaymentService: CardPresentPaymentPreviewService()))
+                                                     cardPresentPaymentService: CardPresentPaymentPreviewService(),
+                                                     orderService: POSOrderService(siteID: Int64.min,
+                                                                                   credentials: Credentials(authToken: "token"))))
```
See other comment about adding an `OrderPreviewService`

joshheald reviewed (approved):
Nice to see it all coming together!

I've left various comments on how it should/could work, feel free to address on this or another PR.

bozidarsevo commented:
> feel free to address on this or another PR.

Will do it here since it is not a major change :)

bozidarsevo committed:
Add POSOrderPreviewService

bozidarsevo committed:
Remove createTestOrder

bozidarsevo committed:
Use DDLogInfo

bozidarsevo committed:
Remove Task

bozidarsevo committed:
Update PointOfSaleDashboardViewModel.swift

bozidarsevo committed:
Update TotalsView.swift

bozidarsevo commented:
> > feel free to address on this or another PR.
> 
> Will do it here since it is not a major change :)

Made some changes, can you take another look @joshheald ?

joshheald commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -138,32 +126,42 @@ private extension TotalsView {
+            Spacer()
+            Text(formattedPrice ?? "-----")
+                .font(Font.system(size: 20))
+                .redacted(reason: formattedPrice == nil ? [.placeholder] : [])
+                .shimmering(active: shimmeringActive)
```
I noticed that after the order is created, the taxes and total stay shimmering even after they show the unredacted detail.

I think two things are happening here:
1. We're not resetting `viewModel.isSyncingOrder` until the payment completes, which is unexpected, but it's because we `await` the `totalsViewWillAppear()` call within syncOrder. We can reset the flag outside the defer to resolve this.
2. However... we probably don't want different conditions for the value to be shimmering vs redacted in this case. Can we just use `formattedPrice == nil` for both?

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -208,49 +264,39 @@ private extension PointOfSaleDashboardViewModel {
+            isSyncingOrder = false
+        }
+        do {
+            isSyncingOrder = true
+            let order = try await orderService.syncOrder(cart: cart,
+                                                              order: order,
+                                                              allProducts: items)
             self.order = order
```
```suggestion
            isSyncingOrder = true
            let order = try await orderService.syncOrder(cart: cart,
                                                              order: order,
                                                              allProducts: items)
            self.order = order
            isSyncingOrder = false
```

If we keep the defer in place, it means that we don't end up forgetting when errors are thrown... but we could repeat the call in the catch block if you prefer.

joshheald committed:
12763 Don’t show POS if credentials aren’t found

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald reviewed (approved):
Looks good, I think we should merge this now.

I found one more thing which I raised about the shimmering of the totals continuing after the order is synced. Feel free to tackle here or on another PR.

joshheald committed:
Revert accidental scheme change

joshheald committed:
12763 make collect payment button connect reader

joshheald committed:
12763 perform payment after order updated

Ensuring that we stop the syncing flag when the order is synced, and only trigger the payment preparation then as well, allows us to be sure we’re collecting the correct amount.
